### DISCUSSION TITLE
unset ui.toggle

### DISCUSSION THREAD

**Discussion by superlou** [zauberzeug/nicegui#1550] [Q&A] [answered]
### Description

I am trying to receive notifications whenever a toggle button is set to one of two values, as well as when it is unset. The basic example is:

```py
from nicegui import ui

def on_update(evt):
    print(evt)

toggle = ui.toggle(["Pass", "Fail"])
toggle.on("update:model-value", on_update)

ui.run()
```

When you click Pass, the event is printed:

```py
GenericEventArguments(sender=<nicegui.elements.toggle.Toggle object at 0x7f8ddafa2ad0>, client=<nicegui.client.Client object at 0x7f8d74137580>, args=[0, {'value': 0, 'label': 'Pass'}])
```

However, when you click Pass a second time to deselect it, there is no event printed. I would expect to have an event fired with a value of None, or some other indication that the toggle has been changed.

**Comment by superlou** [comment]
Um, this is embarrassing. I forgot I have a click and keydown handler that resets the value to None if you click the already selected value. Without that, the idea of getting an `update:model-value` if you click the same button doesn't make any sense. Would being able to unset a toggle be a useful feature? This should move to discussion.

**Comment by rodja** [comment, marked as answer]
You can "unset" a toggle programatically by setting its value to `None`. To "unset" the toggle value if it is clicked again you could try:

```py
current_value = None

def on_click():
    global current_value
    if current_value == toggle.value:
        toggle.value = None
    current_value = toggle.value


toggle = ui.toggle(["Pass", "Fail"])
toggle.on("click", on_click)
```

**Comment by superlou** [comment]
That's what I ended up doing: https://github.com/superlou/nice-datasheet/blob/4aa1dc52f13cfcec54f5a9167a5cc2561b3e025e/step.py#L38-L42. It would be nice to have a way to do it without keeping track of the previous value, but if I can figure out how to extend `ui.toggle`, I could package that up.

**Reply by rodja** [reply]
`ui.toggle` is a class which you can derive your own class from. Python-atypical we picked lower case class names because we want these calls to look more like functions.

**Reply by superlou** [reply]
I didn't realize that. Because of the naming, I thought they were functions that did behind-the-scenes-stuff to make instances of the classes. I'll give that a shot, thanks!

**Comment by falkoschindler** [comment]
For the toggle value to be clearable, Quasar has a `clearable` prop:

```py
ui.toggle(['Pass', 'Fail']).props('clearable')
```

There is, however, a bug which prevents the value from being set to `None`. We'll need to fix it first, but then this would be the way to go.

**Reply by falkoschindler** [reply]
See #1644.
